KAFKA-7433: Introduce broker options in TopicCommand to use AdminClient (KIP-377)

viktorsomogyi opened the pull request:
The PR adds --bootstrap-server and --admin.config options to TopicCommand and implements an alternative, AdminClient based way of topic management.

As testing I've duplicated the existing tests and made them working with the AdminClient options.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

akatona84 commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
 import kafka.server.ConfigType
-import kafka.utils._
+import kafka.utils.Implicits._
+import kafka.utils.{Whitelist, _}
```
could this be `import kafka.utils._`?

smurakozi replied:
```suggestion
import kafka.utils._
```

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+  }
 
+  class CommandTopicPartition(opts: TopicCommandOptions) {
+    val name: String = opts.topic.get
```
Do we always have a topic?

viktorsomogyi replied:
Practically yes. Using the TopicCommand as intended (via command line) would execute checkArgs() which would check if the given argument combination is valid and if we need the topic we have it or not. This member is used only in those code parts where topic is always required, therefore I can safely call `get`. Would you recommend to refactor it so it'd be more obvious?

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+                                   describeConfigs: Boolean)
+
+  class DescribeOptions(opts: TopicCommandOptions) {
+    val describeConfigs: Boolean = !opts.reportUnavailablePartitions && !opts.reportUnderReplicatedPartitions
```
```suggestion
    val describeConfigs = !opts.reportUnavailablePartitions && !opts.reportUnderReplicatedPartitions
```

viktorsomogyi replied:
The scala style guide (https://docs.scala-lang.org/style/types.html) suggests that type annotations should be used if it's not a private member, so I'd keep this as it is unless you insist. :)

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+
+  class DescribeOptions(opts: TopicCommandOptions) {
+    val describeConfigs: Boolean = !opts.reportUnavailablePartitions && !opts.reportUnderReplicatedPartitions
+    val describePartitions: Boolean = !opts.reportOverriddenConfigs
```
```suggestion
    val describePartitions = !opts.reportOverriddenConfigs
```

viktorsomogyi replied:
As per the scala style guide (https://docs.scala-lang.org/style/types.html) I'll keep this as it is.

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+      (!opts.reportUnderReplicatedPartitions && !opts.reportUnavailablePartitions) ||
+        (opts.reportUnderReplicatedPartitions && partitionDesc.isr.size < partitionDesc.assignedReplicas.size) ||
+        (opts.reportUnavailablePartitions && (partitionDesc.leader.isEmpty || !liveBrokers.contains(partitionDesc.leader.get)))
+    }
```
It would be nice to find some descriptive names for the or-ed conditions

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+          case None =>
+        }
+        JAdminClient.create(commandConfig)
+      })
```
I would move the internal block that creates the adminClient to a separate function.

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+          adminZkClient.createTopic(topic.name, topic.partitions.get, topic.replicationFactor, topic.configsToAdd, topic.rackAwareMode)
+        println("Created topic \"%s\".".format(topic.name))
+      } catch  {
+        case e: TopicExistsException => if (!topic.ifTopicDoesntExist()) throw e
```
What's the purpose of this if?

viktorsomogyi replied:
That is a weird logic to handle a conditional operation. :)
So there is an `--if-not-exists` option in TopicCommand which would create the topic if it doesn't exist. Now the implementation will try to create the topic anyways but specifying this flag will suppress the exception. I didn't refactor this too much because it'll be deprecated and didn't want to change its behavior. In `AdminClientTopicService.createTopic` it is implemented properly.

viktorsomogyi replied:
Also looking at the bigger picture I'm not 100% convinced that we'll need the mentioned flag because for more advanced logic people should use the AdminClient directly, so we might even deprecate it.

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+      val topics = getTopics(opts.topic, opts.excludeInternalTopics)
+      for(topic <- topics) {
+        if (zkClient.isTopicMarkedForDeletion(topic))
+          println("%s - marked for deletion".format(topic))
```
```suggestion
          println(s"$topic - marked for deletion")
```

akatona84 commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -295,80 +450,145 @@ object TopicCommand extends Logging {
-    val options = parser.parse(args : _*)
+    private val options = parser.parse(args : _*)
 
     val allTopicLevelOpts: Set[OptionSpec[_]] = Set(alterOpt, createOpt, describeOpt, listOpt, deleteOpt)
```
This could be private as well.

akatona84 reviewed (commented):
generally looks good to me, left a few comments

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+      val topic = new CommandTopicPartition(opts)
+      val topics = getTopics(opts.topic, opts.excludeInternalTopics)
+      if (topics.isEmpty && !opts.ifExists) {
+        throw new IllegalArgumentException("Topic %s does not exist".format(topic.name))
```
Use string interpolation

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -17,248 +17,403 @@
+            val numPartitions = td.partitions().size
+            val replicationFactor = td.partitions().iterator().next().replicas().size
+            val configsAsString = config.entries().asScala.filter(!_.isDefault).map { ce => s"${ce.name()}=${ce.value()}" }.mkString(",")
+            println("Topic:%s\tPartitionCount:%d\tReplicationFactor:%d\tConfigs:%s"
```
Use string interpolation, search for .format :)

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -272,8 +427,8 @@ object TopicCommand extends Logging {
   def parseTopicConfigsToBeDeleted(opts: TopicCommandOptions): Seq[String] = {
-    if (opts.options.has(opts.deleteConfigOpt)) {
-      val configsToBeDeleted = opts.options.valuesOf(opts.deleteConfigOpt).asScala.map(_.trim())
+    if (opts.configsToDelete.isDefined) {
```
You can either remove the if or use get instead of getOrElse

smurakozi commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -295,80 +450,145 @@ object TopicCommand extends Logging {
+    def hasDescribeOption: Boolean = has(describeOpt)
+    def hasDeleteOption: Boolean = has(deleteOpt)
+
+    def zkConnect: Option[String] = if (has(zkConnectOpt)) Some(options.valueOf(zkConnectOpt)) else None
```
Why not introduce a function for this logic? You use it quite many times.

viktorsomogyi committed:
Add new options and helper functions

viktorsomogyi committed:
Introduce TopicService trait and helper classes

viktorsomogyi committed:
Refactor and implement create topic functionality

viktorsomogyi committed:
Add extra test cases for TopicCommandTest

viktorsomogyi committed:
Add tests for TopicCommand with AdminClient

viktorsomogyi committed:
Implement list-topics and refactor get-topics

viktorsomogyi committed:
Implement and refactor alter-topic

viktorsomogyi committed:
Implement and refactor create-topic

viktorsomogyi committed:
Implement and refactor delete-topics

viktorsomogyi committed:
Minor adjustment to topic-list

viktorsomogyi commented:
Updated the review with a nicer rebase, so it's possible to diff in smaller chunks (by commit).

viktorsomogyi commented:
@cmccabe @harshach, pinging you for reviews since you were involved in the discussion / vote of KIP-377. Would you please review this?

viktorsomogyi commented:
@omkreddy, @ijuma, would you please review this?

viktorsomogyi commented:
retest this please

viktorsomogyi committed:
Remove unnecessary apply methods

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,582 @@
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package unit.kafka.admin
```
remove "unit" from package name

viktorsomogyi replied:
Fixed.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,582 @@
+import kafka.common.AdminCommandFailedException
+import kafka.integration.KafkaServerTestHarness
+import kafka.server.{ConfigType, KafkaConfig}
+import kafka.utils.ZkUtils.getDeleteTopicPath
```
ZkUtils is deprecated. We need to use kafka.zk.DeleteTopicsTopicZNode

viktorsomogyi replied:
Fixed.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,582 @@
+    }
+  }
+
+  def assertCheckArgsExitCode(expected: Int, options: TopicCommandOptions) {
```
unused method param?

viktorsomogyi replied:
Fixed it, thanks.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,582 @@
+  private var topicService: AdminClientTopicService = _
+  private var adminClient: AdminClient = _
+
+  def assertExitCode(expected: Int, method: () => Unit) {
```
unused method param?

viktorsomogyi replied:
Fixed it, thanks.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,582 @@
+  @Test
+  def testDescribeReportOverriddenConfigs(): Unit = {
+    val config = "file.delete.delay.ms=1000"
+    val configResource = new ConfigResource(ConfigResource.Type.TOPIC, "testTopic")
```
maybe we can just have variable for topic name

viktorsomogyi replied:
In my patch I've extracted these and use a generated name instead that consists of the test's name and a random alphanumeric string.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandTest.scala:
```diff
@@ -16,19 +16,269 @@
 import kafka.server.ConfigType
-import kafka.admin.TopicCommand.TopicCommandOptions
-import org.apache.kafka.common.errors.TopicExistsException
+import kafka.utils.ZkUtils.getDeleteTopicPath
```
ZkUtils is deprecated. We need to use kafka.zk.DeleteTopicsTopicZNode

viktorsomogyi replied:
Replaced it, thanks.

viktorsomogyi committed:
Replace ZkUtils, use generated topic name in tests

- ZkUtils replaced with DeleteTopicsTopicZNode
- Package name of TopicCommandWithAdminClientTest changed to kafka.admin
- Using generated topic names for tests that consists of the test method's name and a random alphanumeric string

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -42,329 +47,554 @@ object TopicCommand extends Logging {
+          new NewTopic(topic.name, topic.partitions.get, topic.replicationFactor.shortValue())
+
+        val createResult = adminClient.createTopics(Collections.singleton(newTopic))
+        createResult.all().get()
```
since adminClient.createTopics is a asynchronous call, there is a chance adminClient.alterConfigs can fail with topic not exists exception. Can we add waitTillTopicCreation() method?

viktorsomogyi replied:
Hmm, `createResult.all().get()` will block until the future returns, that is the topic has been created or some error happened, so I'm not sure what do you mean here. Should I refactor this logic to a private method?

omkreddy replied:
I was mainly referring to this [comment](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java#L110). But this should not be an issue for alterConfigs.

looks like we can set configs using NewTopics.configs() before calling createTopics(). Additional call to adminClient.alterConfigs may not be required.

viktorsomogyi replied:
Yea, I think it makes much more sense to use `NewTopics.configs` instead of the second call. I've changed that.

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -42,329 +47,554 @@ object TopicCommand extends Logging {
+      val adminZkClient = new AdminZkClient(zkClient)
+      topics.foreach { topic =>
+        val configs = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
+        if(opts.topicConfig.isDefined || opts.configsToDelete.isDefined) {
```
Can we add a test for zk based alter configs?

viktorsomogyi replied:
Done, see `TopicCommandTest.testAlterConfigs`

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -42,329 +47,554 @@ object TopicCommand extends Logging {
                            .describedAs("name=value")
                            .ofType(classOf[String])
-    val deleteConfigOpt = parser.accepts("delete-config", "A topic configuration override to be removed for an existing topic (see the list of configurations under the --config option).")
+    private val deleteConfigOpt = parser.accepts("delete-config", "A topic configuration override to be removed for an existing topic (see the list of configurations under the --config option).")
```
Can we update the description stating that this config only supported with zk option

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -42,329 +47,554 @@ object TopicCommand extends Logging {
-    val nl = System.getProperty("line.separator")
-    val configOpt = parser.accepts("config", "A topic configuration override for the topic being created or altered."  +
+    private val nl = System.getProperty("line.separator")
+    private val configOpt = parser.accepts("config", "A topic configuration override for the topic being created or altered."  +
```
Can we update the description stating that update config is only supported with zk option

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -42,329 +47,554 @@ object TopicCommand extends Logging {
+      if (has(createOpt) && !has(replicaAssignmentOpt))
+        CommandLineUtils.checkRequiredArgs(parser, options, partitionsOpt, replicationFactorOpt)
+      if (has(alterOpt))
+        CommandLineUtils.checkRequiredArgs(parser, options, partitionsOpt)
```
with this we can not alter/delete configs using zk option

viktorsomogyi replied:
Thanks, didn't notice this. Will modify this to `if (has(bootstrapServerOpt) && has(alterOpt))`.

omkreddy commented on core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala:
```diff
@@ -0,0 +1,564 @@
+  @Test
+  def testInvalidConfigOptWithBootstrapServers(): Unit = {
+    assertCheckArgsExitCode(1,
+      new TopicCommandOptions(Array("--bootstrap-server", brokerList ,"--alter", "--topic", testTopicName, "--config", "cleanup.policy=compact")))
```
This will fail with below error. But error should be " --config"  is not allowed.
``` 
>>./kafka-topics.sh --bootstrap-server localhost:9092  --alter --topic foo --config cleanup.policy=compact
Missing required argument "[partitions]"

//Below zk based command should work

>>./kafka-topics.sh --zookeeper localhost:2181 --alter --topic foo --config cleanup.policy=compact
Missing required argument "[partitions]"
```

viktorsomogyi replied:
Well, in this case I think I'm gonna fix the test as the partitions opt should be required for the `alter` option in this case since `alter` is reduced to the functionality of changing the partition count.

omkreddy reviewed (commented):
@viktorsomogyi Thanks for the updated PR. Left few more comments.

omkreddy commented:
Can we also update the docs? can be done in another PR.

viktorsomogyi committed:
Shouldn't support --if-exists and --if-not-exists with --bootstrap-server

viktorsomogyi committed:
Add extra test for ZK alter config

viktorsomogyi committed:
Use generatd topic names in TopicCommandTest

viktorsomogyi committed:
Add not supported notes to command line descriptions

viktorsomogyi committed:
Fix argument checks

viktorsomogyi commented:
@omkreddy I've created KAFKA-7804 for updating the docs (both examples and notable changes) and  KAFKA-7805 for updating the system tests with this new option. I'll also work on these during the following days.

viktorsomogyi committed:
Use NewTopic.configs instead of a second api call

viktorsomogyi committed:
Add extra case to testAlterConfigs

viktorsomogyi commented:
retest this please

viktorsomogyi commented:
retest this please

viktorsomogyi committed:
Option check should allow configOpt with createOpt and bootstrapServerOpt

viktorsomogyi committed:
Add wait for topic creation in tests as it's not transactional

viktorsomogyi committed:
Fix ReassignPartitionsCommandArgsTest typos due to CommandLineUtils change

viktorsomogyi commented:
@omkreddy would you please review this again? I've eliminated the flaky tests and ran ducktapes with the updated kafka-topics (not the full set yet though) to increase the confidence of this patch.

omkreddy commented:
@viktorsomogyi sorry for thr delay. I will take another look in next couple of days.

viktorsomogyi commented:
@omkreddy sure, thanks for updating me.

omkreddy commented:
@viktorsomogyi  Thanks for the Updated PR. LGTM. We should be able to merge this by 2.2.0 feature freeze date (Jan 31st)

@junrao can take a look at this when you have time?

KevinLiLu commented:
@omkreddy @junrao any update regarding this?

viktorsomogyi commented:
@junrao @hachikuji would you please review this so it can be merged by the feature freeze today?

hachikuji commented:
@omkreddy I did a high-level pass and the patch looks solid to me. If you are happy with it, I'd suggest we go ahead and merge. It seems like the new behavior is pretty well isolated, so I don't have any major concerns.

omkreddy commented:
@viktorsomogyi  Thanks for the PR. LGTM.  merging to trunk.  We should fix if any issues in follow-up PRs.

omkreddy commented on core/src/main/scala/kafka/admin/TopicCommand.scala:
```diff
@@ -259,129 +410,210 @@ object TopicCommand extends Logging {
                                              "The following is a list of valid configurations: " + nl + LogConfig.configNames.map("\t" + _).mkString(nl) + nl +
-                                             "See the Kafka documentation for full details on the topic configs.")
+                                             "See the Kafka documentation for full details on the topic configs." +
+                                             "Not supported with the --bootstrap-server option.")
```
@viktorsomogyi  looks like this is incorrect. We can use  bootstrap-server + create + config option combination. Can we update this as part of #6118?

viktorsomogyi replied:
Thanks, I'll update that.

viktorsomogyi commented:
@hachikuji @omkreddy @smurakozi @akatona84 many thanks for the code review, the effort that you put in :)
